sdk/core/core-client/src/utils.ts

```diff
@@ -82,30 +107,38 @@ export function flattenResponse(
           arrayResponse[key] = parsedHeaders[key];
         }
       }
-      return arrayResponse;
+      return isNull && arrayResponse === emptyArray ? null : arrayResponse;
     }
 
     if (typeName === "Composite" || typeName === "Dictionary") {
-      return {
+      const response = {
         ...parsedHeaders,
         ...fullResponse.parsedBody
       };
+      return returnNullResponse(isNullable, response, response);
     }
   }
 
+  const response = {
+    ...parsedHeaders,
+    ...fullResponse.parsedBody
+  };
+
   if (
     bodyMapper ||
     fullResponse.request.method === "HEAD" ||
     isPrimitiveType(fullResponse.parsedBody)
   ) {
-    return {
-      ...parsedHeaders,
-      body: fullResponse.parsedBody
-    };
+    return returnNullResponse(
+      isNullable,
+      response,
+      {
+        ...parsedHeaders,
+        body: fullResponse.parsedBody
+      },
+      true /** is primitive */
```

Comment:
can the argument to this function actually be an object with named properties? That would give better semantic meaning to these boolean parameters

Reply:
I see, I simplified the inputs to the function and put them in an object here: https://github.com/Azure/azure-sdk-for-js/pull/14366/commits/dd689fadc4575fe49951599a9ef276dc94a2a8e3.